### ISSUE TITLE
Message cancellation

### ISSUE THREAD

**Issue by robshep** [celrenheit/sandglass#19]
Hi,

I'm very interested in this project and would like to find out how much work would be involved in adding functionality so that future messages can be cancelled.

Assuming the storage key was retained by another component.

I'd have a go at implementing this given a few pointers and i'll be happy to create a Java client if I can satisfy my requirements for this project.

Cheers

**Comment by celrenheit** [comment]
Hi,

It depends on what you mean by cancelation.
If it is only for a specific consumer group, you can already do it by marking the message as ACKed. It won't be delivered.
Otherwise, if you mean to delete the message forever even for future consumer groups, then that it is not yet implemented. I have a few idea on how to implement this.

**Comment by robshep** [comment]
Yes, it would be a producer with new knowledge that needs to prevent a previous future message from being delivered. 

In pseudo-world......

Client > Schedule Appointment for 15:00hrs
Producer > create message {
  Offset: 1430hrs-now() 
  Topic: "sms-reminders" 
  Value: JSON.stringify( {c_id: 1, apt_id: 5, apt_time: 1500 } )
}

Client > Cancel Appointment
...
[tick tock, it's 14:30 hrs]
Consumer > send reminder to client 1 about appointment 5 at 15:00 #fail!

I'd be interested to hear your ideas and seeing if it's something I can help with.

**Comment by celrenheit** [comment]
For the given example, we can say that this using a consumer group named `sms_senders` responsible for sending the SMSes. If you ACK the message that should be delivered at 14:30, it will not be delivered when it will be 14:30.
However, if now you have another consumer group `sms_stats`that does some stats on the reminders, then the message will be delivered since it is not ACKed for this consumer group.

To avoid a message from being consumed ever again from any consumer group, I have to think more about it.

**Comment by robshep** [comment]
Thank you for the explanation. 
This sounds fine for my use case. 

Do you have a code snippet, or a bit in the codebase, that can help me discover how to ACK a previously queued message?

Thanks

**Comment by celrenheit** [comment]
You can use this method when using the go client: https://godoc.org/github.com/celrenheit/sandglass-client/go/sg#Consumer.Acknowledge
Underneath, it uses this grpc method: https://github.com/celrenheit/sandglass-grpc/blob/master/sgproto/sandglass.proto#L206

But please be careful cause the soon to be released v0.2, we will break many things
